sw/device/lib/base/hardened.h

```diff
@@ -304,4 +304,158 @@ inline void barrier32(uint32_t val) { asm volatile("" ::"r"(val)); }
  */
 inline void barrierw(uintptr_t val) { asm volatile("" ::"r"(val)); }
 
+/**
+ * A constant-time, 32-bit boolean value.
+ *
+ * Values of this type MUST be either all zero bits or all one bits.
+ */
+typedef uint32_t ct_bool32_t;
+
+// NOTE: the functions below are taken from BoringSSL, which includes proofs of
+// correctness by Adam Langley.
+// https://boringssl.googlesource.com/boringssl/+/refs/heads/master/crypto/internal.h#262
+
+/**
+ * Performs constant-time signed comparison to zero.
+ *
+ * Returns whether `a < 0`, as a constant-time boolean.
+ *
+ * @return `a < 0`.
+ */
+inline ct_bool32_t ct_sltz32(int32_t a) {
+  // Arithmetic left shift!
+  return (uint32_t)(a >> 31);
+}
+
+/**
+ * Performs constant-time unsigned ascending comparison.
+ *
+ * Returns `a < b` as a constant-time boolean.
+ *
+ * @return `a < b`.
+ */
+inline ct_bool32_t ct_sltu32(uint32_t a, uint32_t b) {
```

Comment:
Are you planning to use inline assembly for these at some point in the future? Seems like these functions map very nicely onto RISC-V instructions but I doubt the compiler will be able to figure that out.

Reply:
Hmm... interesting thought. I had not planned to do that but it might actually be a good idea, for two reasons:
- It gives us absolute control over the implementation, so we can prevent any rude instruction scheduling effects.
- It blocks strength reduction, which is a real concern if the compiler can deduce that `ct_bool32_t` is always all ones or all zeroes.

Reply:
I'd leave it as an open issue. In the short term we can write a functional test case.  There is an instruction simulator we use for DV which we could integrate here for unittests. Probably a good idea to discuss options.